File: rust/p0287_find_duplicate_number/src/binary_search.rs
pub struct Solution;

impl Solution {
    pub fn find_duplicate(nums: Vec<i32>) -> i32 {
        let (mut left, mut right) = (1, nums.len() as i32 - 1);
        while left < right {
            let mid = left + (right - left) / 2;
            let count = nums.iter().filter(|&&x| x <= mid).count();

            if count > mid as usize {
                right = mid
            } else {
                left = mid + 1
            }
        }
        left
    }
}

File: csharp/LeetCode/P0287_FindDuplicateNumber/solutions/BinarySearch.cs
namespace P0287_FindDuplicateNumber.BinarySearch;

public class Solution
{
    // Binary search implementation
    public int FindDuplicate(int[] nums)
    {
        int left = 1, right = nums.Length - 1;

        while (left < right)
        {
            int mid = left + (right - left) / 2;  // Avoid integer overflow
            // Count numbers <= mid. If count > mid, then the duplicate number is in [1, mid]
            int count = nums.Count(num => num <= mid);
            if (count > mid)
                right = mid;
            else
                left = mid + 1;
        }
        return left;
    }
}
